src/CBLURLEndpointListener.cc

```diff
@@ -0,0 +1,216 @@
+
+                slice usr = decoded.upTo(colon);
+                slice pwd = decoded.from(colon + 1);
+
+                auto me = reinterpret_cast<CBLURLEndpointListener*>(context);
+                Assert(me->_c4listener == listener);
+                return me->_conf.authenticator->pswCallback(me->_conf.context, usr, pwd);
+            };
+            c4config.callbackContext = this;
+        }
+    }
+
+    std::unique_ptr<C4Listener> c4listener{new C4Listener(c4config)};
+    auto ret = [&](bool succ) -> bool {
+        if (succ) {
+            _c4listener = c4listener.release();
+        }
+        return succ;
+    };
+
+    CBLDatabase* cblDb = _conf.collections[0]->database();
+    bool succ = true;
+    cblDb->c4db()->useLocked([&](C4Database* db) {
+        slice dbname = db->getName();
+        if ( (succ = c4listener->shareDB(dbname, db)) ) {
+            for (unsigned i = 0; i < _conf.collectionCount; ++i) {
+                _conf.collections[i]->useLocked([&](C4Collection* coll) {
+                    succ = c4listener->shareCollection(dbname, coll);
+                });
+                if (!succ) break;
+            }
+        }
+    });
+
+    return ret(succ);
+}
+
+void CBLURLEndpointListener::stop() {
+    std::scoped_lock lock(_mutex);
+    if (_c4listener) {
+        delete _c4listener;
+        _c4listener = nullptr;
+    }
+}
+
+std::string CBLURLEndpointListener::dumpConfig() {
+    return {};
+}
+
+CBLTLSIdentity* CBLURLEndpointListener::effectiveTLSIdentity(bool persistent) {
+    if (_conf.disableTLS)
+        return nullptr;
+
+    if (!_effectiveTLSIdentity)
+        _effectiveTLSIdentity = _conf.tlsIdentity ? _conf.tlsIdentity : anonymousTLSIdentity(persistent).get();
+
+    return _effectiveTLSIdentity;
+}
+
+Retained<CBLTLSIdentity> CBLURLEndpointListener::anonymousTLSIdentity(bool persistent) {
+    Retained<CBLTLSIdentity> identity;
+    alloc_slice label;
+    std::unique_ptr<CBLKeyPair, void(*)(CBLKeyPair*)> keypair{
+        nullptr,
+        [](CBLKeyPair* k) {
+            CBLKeyPair_Release(k);
+        }
+    };
+    if (persistent) {
+#if !defined(__linux__) && !defined(__ANDROID__)
+        label = labelForAnonymousTLSIdentity();
+        if (!label) return nullptr;
+
+        identity = CBLTLSIdentity::IdentityWithLabel(label);
+        C4Error error{};
+        if (identity) {
+            CBL_Log(kCBLLogDomainListener, kCBLLogVerbose, "Found anonymous identity by label = '%.*s'", (int)label.size, (char*)label.buf);
+
+            CBLTimestamp expire = identity->expiration();
+            if (expire/1000 > kCBLAnonymousIdentityMinValidTimeAllowed) {
+                return identity;
+            }
+
+            CBL_Log(kCBLLogDomainListener, kCBLLogVerbose, "Delete anonymous identity of label = '%.*s' (expiration = %ld)", (int)label.size, (char*)label.buf, (long)(expire/1000));
+
+            CBLTLSIdentity_DeleteIdentityWithLabel(label, cbl_internal::external(&error));
```

Comment:
This is C++ level so it shouldn't use the public C level api. Use  CBLTLSIdentity::DeleteIdentityWithLabel instead.

Reply:
okay. Let me fix it